### PR TITLE
Accept functions to flow_run_name and task_run_name settings

### PR DESCRIPTION
**Pull request by rsampaths16**
Currently flow_run_name and task_run_name settings only accept template strings to generate new name. This PR enhances this feature by accept a function that expects input parameters dict and returns a string. This should ideally allow injecting flow run context and task run context too.

#### Flow run name setting
 ```python
import datetime
from prefect import flow

def generate_flow_run_name(params):
    date = datetime.datetime.utcnow()
    name = "anon" if "name" not in params else str(params["name"])

    return f"{name}-on-{date:%A}"

@flow(flow_run_name=generate_flow_run_name)
def my_flow(name: str):
    pass

# creates a flow run called 'marvin-on-Thursday'
my_flow(name="marvin")
```

#### Task run name setting
```python
import datetime
from prefect import flow, task

def generate_task_name(params):
    name = "anon" if "name" not in params else params["name"]
    date = datetime.datetime.utcnow()

    return f"hello-{name}-on-{date:%A}")

@task(name="My Example Task", 
      description="An example task for a tutorial.",
      task_run_name=generate_task_name)
def my_task(name):
    pass

@flow
def my_flow():
    # creates a run with a name like "hello-marvin-on-Thursday"
    my_task(name="marvin")
```

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [X] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  

closes https://github.com/PrefectHQ/prefect/issues/8895

### CODE REVIEW

File: src/prefect/engine.py
Comment (rsampaths16) [inline src/prefect/engine.py:622]: **WARNING**: The parameters is sent directly to the `flow_run_name` function which could leave it vulnerable to mutation. We could try to send a shallow copy of the parameters dict, but still the individual args themselves could be mutated ( similar to how it's vulnerable in `str.format(**parameters)` ).

I'll leave the call on what to do here to the Prefect team.

File: src/prefect/engine.py
Comment (rsampaths16) [inline src/prefect/engine.py:628]: In both `flow_run_name` and `task_run_name` I'm raising exceptions when they are neither `str` nor `callable`. Should this assertion be removed? If not should an assertion be added in the actual `Flow` and `Task` classes too?

File: src/prefect/engine.py
Comment (madkinsz) [reply src/prefect/engine.py:622]: I think that's okay with me for now 👍

File: src/prefect/engine.py
Comment (madkinsz) [reply src/prefect/engine.py:628]: Raising errors at the flow/task class level is better — then we will help people do the right thing when writing their task / flow rather than at runtime.

**Comment by rsampaths16** [comment]
> Thanks for contributing!
> 
> Before we dive into a review of the implementation, let's chat about the interface a little. I'm thinking that since I've added `parameters` to the `prefect.runtime` module perhaps this should be `Callable[[], str]` and you should just pull data from the runtime as needed. What do you think?

I believe this would be a good approach. We won't force a people to accept parameters if they don't want to use it, and since they'd be available in `prefect.runtime` those who want them are not blocked.

File: src/prefect/engine.py
Comment (rsampaths16) [reply src/prefect/engine.py:628]: I'll make changes to raise `TypeError`'s in `Flow` & `Task` class level. I'm also thinking to leave the `RuntimeError`'s here. Please confirm if the same if acceptable.

**Comment by madkinsz** [comment]
@rsampaths16 sounds good to me!

**Comment by rsampaths16** [comment]
Hey @madkinsz during testing I've realized that `prefect.runtime.flow_run` tries to access the parent's `flow_run_context`, making the current flows parameters inaccessible to it. I think we'll have to make the function accept parameters as a kwargs dictionary. What are your thoughts on this?

P.S: PR seems to be accidentally closed as I've force pushed a rebase. Will re-open again when I've completed the implementation.

**Comment by rsampaths16** [comment]
Hey @madkinsz - regarding https://github.com/PrefectHQ/prefect/issues/8895#issuecomment-1483060367 what would be the plan to address this? Since the `run_name` functions are being called before we are inside the context of the `flow/task`? I'm assuming other functions that are accepted wouldn't have context too.

File: src/prefect/engine.py
Comment (madkinsz) [reply src/prefect/engine.py]: It's okay here as well, but we'll probably want to move this into a separate function since it's kind of long now.

**Comment by rsampaths16** [comment]
Hi @madkinsz !!

https://github.com/PrefectHQ/prefect/blob/c0828fa478f242610c4fe8dac87c05e364cb5094/src/prefect/engine.py#L1520-L1530


Upon going through the code above, is my understanding correct that since the `task_run_name` function will be called within `task_run_context` context, it'll be able to access the `prefect.runtime` parameters?

If yes, I'll try out accessing `prefect.runtime` from the `task_run_name` and check if a similar pattern works for flow too. Maybe that'd be a separate pull-request.

Do let me know if I'm looking at the wrong pattern here.

**Comment by rsampaths16** [comment]
Awesome @madkinsz! I'll test this in the morning see if I can raise a new revision where functions don't accept parameters as discussed earlier. Thanks.

**Comment by rsampaths16** [comment]
Hey @madkinsz I've made the discussed changes.
The summary is:
1. refactored few tests in flows and tasks by grouping them together
2. add flow_run_name and task_run_name function support: they don't accept any arguments.
3. move flow_run_name generation within the flow run context: so it can access parameters from `prefect.runtime.flow_run.parameters`

For more detailed info you can check-out the commit messages. :)

**Comment by rsampaths16** [comment]
Hey @madkinsz - bumping again incase you've missed this;

File: docs/tutorials/flow-task-config.md
Comment (madkinsz) [inline docs/tutorials/flow-task-config.md:97]:
```suggestion
Additionally this setting also accepts a function that returns a string for the flow run name:
```

File: docs/tutorials/flow-task-config.md
Comment (madkinsz) [inline docs/tutorials/flow-task-config.md:115]: Can we add a blurb here that says something like

> If you need access to information about the flow, use the `prefect.runtime` module. For example:

Then include an example that pulls the name of the flow and a parameter from the runtime?

File: docs/tutorials/flow-task-config.md
Comment (madkinsz) [inline docs/tutorials/flow-task-config.md:191]: Same as above re an additional example but with the task run runtime.

File: src/prefect/engine.py
Comment (madkinsz) [inline src/prefect/engine.py:642]: This will update the name on each retry, right? We'll want to avoid setting the name more than once.

File: src/prefect/flows.py
Comment (madkinsz) [inline src/prefect/flows.py:155]: Generally we want to give more information to help debug
```suggestion
                raise TypeError(f"Expected string or callable for 'flow_run_name'; got {type(flow_run_name).__name__}.")
```

File: src/prefect/tasks.py
Comment (madkinsz) [inline src/prefect/tasks.py:220]: See flow run name error comment

File: src/prefect/engine.py
Comment (madkinsz) [inline src/prefect/engine.py:2085]: We probably want something more verbose for debugging

```suggestion
            raise RuntimeError(f"Callable {task.task_run_name} for 'task_run_name' returned type {type(task_run_name).__name__} but a string is required.")
```

Similar for the other error messages here

File: src/prefect/engine.py
Comment (madkinsz) [inline src/prefect/engine.py:2072]: Note all of these `RuntimeError`s can be `TypeError`s

**Comment by madkinsz** [comment]
Thanks again for contributing @rsampaths16! This looks great. I have some minor comments regarding UX when throwing exceptions.

File: src/prefect/engine.py
Comment (rsampaths16) [reply src/prefect/engine.py:2073]: Hey, I'll convert `RuntimeError`s to `TypeError`s where we check the type of the generating function/string. I'm not sure `TypeError` should be raised for the return value of the function.

Comment (rsampaths16) [reply src/prefect/engine.py:2073]: Just checked that `ValueError` is being raised on invalid return values in the code-base. Since the practice is established, I'll raise `TypeError` for the non string return value.

File: src/prefect/engine.py
Comment (rsampaths16) [reply src/prefect/engine.py:642]: This is updating on each retry. Made changes to avoid setting it more than once. Used the same pattern as implemented in task_run_name. Have also added relevant tests.

File: docs/tutorials/flow-task-config.md
Comment (rsampaths16) [reply docs/tutorials/flow-task-config.md:97]: done

File: docs/tutorials/flow-task-config.md
Comment (rsampaths16) [reply docs/tutorials/flow-task-config.md:115]: done

File: docs/tutorials/flow-task-config.md
Comment (rsampaths16) [reply docs/tutorials/flow-task-config.md:170]: done

File: docs/tutorials/flow-task-config.md
Comment (rsampaths16) [reply docs/tutorials/flow-task-config.md:191]: done

File: src/prefect/engine.py
Comment (rsampaths16) [reply src/prefect/engine.py:642]: done

File: src/prefect/engine.py
Comment (rsampaths16) [reply src/prefect/engine.py:2073]: done

File: src/prefect/engine.py
Comment (rsampaths16) [reply src/prefect/engine.py:2086]: done

File: src/prefect/flows.py
Comment (rsampaths16) [reply src/prefect/flows.py:155]: done

File: src/prefect/tasks.py
Comment (rsampaths16) [reply src/prefect/tasks.py:220]: done

File: tests/test_flows.py
Comment (rsampaths16) [reply tests/test_flows.py:177]: removed

File: tests/test_flows.py
Comment (rsampaths16) [reply tests/test_flows.py:178]: dropped

**Comment by rsampaths16** [comment]
Hey @madkinsz - here are the changes since the last review:
* added documentation for using `prefect.runtime` within `flow_run_name` and `task_run_name` callable
* addressed nitpicks in documentation
* made exception messages more verbose
* changed `RuntimeError`s to `TypeError`s in `prefect.engine`
* addressed nitpicks in tests
* fixed issue where `flow_run_name` is reset on retries. ( followed the same pattern used in `task_run_name`. added missing tests too ).

do let me know your thoughts on it

**Comment by rsampaths16** [comment]
Hi @madkinsz - I've verified that the tests failing aren't related to the current pr. Could you please check and unblock any workflows that could've been blocked by this. Thanks.

ref: https://github.com/PrefectHQ/prefect/pull/9239

**Comment by rsampaths16** [comment]
Hi @madkinsz - thanks for unblocking the tests via https://github.com/PrefectHQ/prefect/pull/9241!!

**Review by madkinsz — APPROVED** [review]
Thanks for your patience!

**Comment by rsampaths16** [comment]
Awesome :)

**Comment by batuhanfaik** [comment]
Current implementation :)

```python
import datetime
from prefect import flow

def generate_flow_run_name():
    flow_run = prefect.runtime.flow_run
    date = datetime.datetime.utcnow()
    name = "anon" if not flow_run.name else str(flow_run.name)
    return f"{name}-on-{date:%A}"

@flow(flow_run_name=generate_flow_run_name)
def my_flow(name: str):
   pass

# creates a flow run called 'marvin-on-Thursday'
my_flow(name="marvin")
```

**Pull request merged**
